includes/Story_Post_Type.php

```diff
@@ -162,20 +171,103 @@ protected function get_post_type_icon() {
+			$current_post = get_post();
+			if ( self::AMP_VALIDATED_URL_POST_TYPE === $current_screen->post_type && $current_post instanceof WP_Post && $current_post->post_type === $current_screen->post_type ) {
+				$validated_url_post_type = $this->get_validated_url_post_type( $current_post->ID );
+				if ( $validated_url_post_type ) {
+					return $validated_url_post_type;
+				}
+			}
+			if ( $current_screen->post_type ) {
+				return $current_screen->post_type;
+			}
+		} elseif ( isset( $_SERVER['REQUEST_URI'] ) && false !== strpos( (string) wp_unslash( $_SERVER['REQUEST_URI'] ), '/web-stories/v1/web-story/' ) ) { // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+			return self::POST_TYPE_SLUG;
+		}
+		return null;
+	}
+
+	/**
+	 * Get the singular post type which is the queried object for the given validated URL post.
 	 *
-	 * @param bool    $skipped Should this post type be skipped.
-	 * @param WP_Post $post Post object.
+	 * @param int $post_id Post ID for Validated URL Post.
+	 * @return string|null Post type or null if validated URL is not for a singular post.
+	 */
+	protected function get_validated_url_post_type( $post_id ) {
+		if ( empty( $post_id ) ) {
+			return null;
+		}
+
+		$post = get_post( $post_id );
+		if ( ! ( $post instanceof WP_Post ) ) {
+			return null;
+		}
+
+		if ( self::AMP_VALIDATED_URL_POST_TYPE !== $post->post_type ) {
+			return null;
+		}
+
+		$queried_object = get_post_meta( $post->ID, '_amp_queried_object', true );
+		if ( isset( $queried_object['id'], $queried_object['type'] ) && 'post' === $queried_object['type'] ) {
+			$post_type = get_post_type( $queried_object['id'] );
+			if ( $post_type ) {
+				return $post_type;
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Filter AMP options to force Standard mode (AMP-first) when a web story is being requested.
 	 *
-	 * @return bool
+	 * @param array $options Options.
+	 * @return array Filtered options.
 	 */
-	public function skip_amp( $skipped, $post ) {
-		if ( self::POST_TYPE_SLUG === get_post_type( $post ) ) {
-			$skipped = true;
+	public function filter_amp_options( $options ) {
+		if ( $this->get_request_post_type() === self::POST_TYPE_SLUG ) {
+			$options['theme_support']          = 'standard';
+			$options['supported_post_types'][] = self::POST_TYPE_SLUG;
+			$options['supported_templates'][]  = 'is_singular';
+		}
+		return $options;
+	}
+
+	/**
+	 * Filter the post types which are supportable.
+	 *
+	 * Remove web-stories from the list unless the currently requested post type is for a web-story. This is done in
+	 * order to hide stories from the list of supportable post types on the AMP Settings screen.
+	 *
+	 * @param string[] $post_types Post types.
+	 * @return array Supportable post types.
+	 */
+	public function filter_supportable_post_types( $post_types ) {
+		if ( $this->get_request_post_type() === self::POST_TYPE_SLUG ) {
+			$post_types = array_merge( $post_types, [ self::POST_TYPE_SLUG ] );
```

Comment:
```suggestion
			$post_types[] = self::POST_TYPE_SLUG;
```

Reply:
That will have the same effect, yes. But `array_merge()` has a small advantage I think because it allows the code to have more symmetry with the `else` condition where `array_diff()` does the opposite.